flink-core/src/main/java/org/apache/flink/util/ResourceGuard.java

```diff
@@ -0,0 +1,130 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.util;
+
+import java.io.IOException;
+import java.io.Serializable;
+
+/**
+ * This class is a guard for shared resources with the following invariants. The resource can be acquired by multiple
+ * clients in parallel through the {@link #acquireResource()} call. Clients must release the resource after usage
+ * with a matching call to {@link #releaseResource()}. The resource can only be disposed once the guard is closed, but
+ * the guard can only be closed once all clients that acquired the resource have released it. Before this is happened,
+ * the call to {@link #close()} will block until the condition is triggered. After the guard is closed, calls to
+ * {@link #acquireResource()} will fail with exception. Notice that, obviously clients are responsible to release the
+ * resource after usage. All clients are considered equal, i.e. there is only a global count maintained how many
+ * times the resource was acquired but not by whom.
+ */
+public class ResourceGuard implements AutoCloseable, Serializable {
+
+	private static final long serialVersionUID = 1L;
+
+	/**
+	 * Introduced additional checks of invariants.
+	 */
+	private static final boolean STRICT_CHECKS = true;
+
+	/**
+	 * The object that serves as lock for count and the closed-flag.
+	 */
+	private final SerializableObject lock;
+
+	/**
+	 * Number of clients that have ongoing access to the resource.
+	 */
+	private int clientCount;
+
+	/**
+	 * This flag indicated if it is still possible to acquire access to the resource.
+	 */
+	private volatile boolean closed;
+
+	public ResourceGuard() {
+		this.lock = new SerializableObject();
+		this.clientCount = 0;
+		this.closed = false;
+	}
+
+	/**
+	 * Acquired access from one new client for the guarded resource.
+	 *
+	 * @throws IOException when the resource guard is already closed.
+	 */
+	public void acquireResource() throws IOException {
+
+		synchronized (lock) {
+
+			if (closed) {
+				throw new IOException("Resource guard was already closed.");
+			}
+
+			++clientCount;
+		}
+	}
+
+	/**
+	 * Releases access for one client of the guarded resource. This method must only be called after a matching call to
+	 * {@link #acquireResource()}.
+	 */
+	public void releaseResource() {
+
+		synchronized (lock) {
+
+			if (STRICT_CHECKS && clientCount <= 0) {
+				throw new IllegalStateException("Resource guard was released more times than it has been acquired!");
+			}
+
+			--clientCount;
+
+			if (closed && clientCount == 0) {
+				lock.notifyAll();
+			}
+
+		}
+	}
+
+	/**
+	 * Closed the resource guard. This method will block until all calls to {@link #acquireResource()} have seen their
+	 * matching call to {@link #releaseResource()}.
+	 */
+	@Override
+	public void close() {
+
+		synchronized (lock) {
+
+			closed = true;
+
+			while (clientCount > 0) {
+
+				try {
+					lock.wait();
+				} catch (InterruptedException e) {
+					Thread.interrupted();
```

Comment:
This line caught my eye.
```
public static boolean interrupted()
```
> Tests whether the current thread has been interrupted. [...]

The boolean result is not used and when `InterruptedException` is thrown, the interrupt flag is cleared, so the method should return `false` here.

Did you intend to call `Thread.currentThread().interrupt()` and exit the `while` loop prematurely?

Reply:
No, exactly this method was my intention. The API is very bad, but I am not interested in the result (because it is already clear here). What I am interested in is the side effect to clear the interrupted flag. If this is not cleared, future calls to `wait` trigger the interrupted exception again.

Reply:
Just double checked, seems like I remembered incorrectly and this call is not needed here because the flag is reset already by the triggered exception.